### ISSUE TITLE
Suggestions for xlsx_from_polarsdf utility

### ISSUE THREAD

**Issue by bertiewooster** [mpm896/HTSCluster#1]
In [polars_xlsx.py](https://github.com/mpm896/HTSCluster/blob/main/HTSCluster/utils/polars_xlsx.py):

Is there a way to write to multiple cells at once with xlsxwriter? That would provide an easy way to get away from having to iterate over each cell in the dataframe, which would presumably improve performance--avoiding [`df.iter_rows()`](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.iter_rows.html) is a good idea if possible.

As a more limited change, to [`enumerate`](https://docs.python.org/3.8/library/functions.html#enumerate) you can provide a `start` value of 1 so you don't have to manually add one to `col_idx` each iteration.

**Comment by mpm896** [comment]
Hi Jeremy, I've been reading through the `xlsxwriter` docs and haven't been able to find anything on writing multiple cells at the same time. I imagine this could be possible with `asyncio` or threading, but I really don't know much about these yet. I've noticed that writing an excel sheet of ~350 compounds with images takes a few seconds, so it definitely isn't efficient.

To be honest, I wrote up this utility function in a pinch and changed as little as possible from the original `PandasTools.SaveXlsxFromFrame` to make it compatible with `Polars`, so since this function was originally written for `Pandas` frames I'm sure there's a bit that can be done to optimize it. `Polars` has a `write_excel()` method that relies on `xlsxwriter` and so I could look into implementing it this way

**Comment by mpm896** [comment]
I've re-written `xlsx_from_polarsdf` in [polars_xlsx.py](https://github.com/mpm896/HTSCluster/blob/main/HTSCluster/utils/polars_xlsx.py) and see a 1.4x speedup (numbers in the README). However the rate limiting step is iterating over all the images and writing them one by one - this hasn't been implemented in `Polars` and there is only the option to write one image at a time using `xlsxwriter`. This is also creates a lot of memory pressure as the DataFrames get larger (for the test library of 43,000 compounds, my M2 Mac was really working hard).

This could be a good opportunity to write a rust implementation with python bindings of writing images to an xlsx format for improved speed and memory efficiency, or modify the `Polars.DataFrame.write_excel()` method to write images

**Comment by bertiewooster** [comment]
Yes, I did the same exercise and came to much the same conclusion--Polars would be somewhat, but not massively, faster than Pandas because processing images is the rate-limiting step.
